File: labs/control-ui/api-shadow/masque_datagram_tunnel.ts
export class MasqueDatagramTunnel {
  public contextId: number;
  constructor(contextId: number = 0) {
    this.contextId = contextId;
  }

  encodeDatagram(payload: Uint8Array): Uint8Array {
    const varintBytes = this.encodeVarint(this.contextId);
    const frame = new Uint8Array(varintBytes.length + payload.length);
    frame.set(varintBytes, 0);
    frame.set(payload, varintBytes.length);
    return frame;
  }

  decodeDatagram(data: Uint8Array): { contextId: number; payload: Uint8Array } | null {
    if (data.length === 0) return null;
    const res = this.decodeVarint(data);
    if (!res) return null;
    const { val, len } = res;
    return {
      contextId: val,
      payload: data.slice(len),
    };
  }

  private encodeVarint(val: number): Uint8Array {
    if (val < 64) {
      return new Uint8Array([val]);
    } else if (val < 16384) {
      const v = 0x4000 | (val & 0x3fff);
      return new Uint8Array([(v >> 8) & 0xff, v & 0xff]);
    } else {
      const v = (0x80000000 | (val & 0x3fffffff)) >>> 0;
      const buf = new Uint8Array(4);
      const view = new DataView(buf.buffer);
      view.setUint32(0, v);
      return buf;
    }
  }

  private decodeVarint(data: Uint8Array): { val: number; len: number } | null {
    if (data.length === 0) return null;
    const first = data[0]!;
    const prefix = first >> 6;
    if (prefix === 0) {
      return { val: first & 0x3f, len: 1 };
    } else if (prefix === 1) {
      if (data.length < 2) return null;
      const val = ((first & 0x3f) << 8) | data[1]!;
      return { val, len: 2 };
    } else if (prefix === 2) {
      if (data.length < 4) return null;
      const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
      const val = view.getUint32(0) & 0x3fffffff;
      return { val, len: 4 };
    }
    return null;
  }
}

File: labs/mobile/donor-archive/tunnel/MasqueDatagramTunnel.kt
package com.luminet.android.tunnel

import java.nio.ByteBuffer

class MasqueDatagramTunnel(val contextId: Long = 0) {

    fun encodeDatagram(payload: ByteArray): ByteArray {
        val varintBytes = encodeVarint(contextId)
        val buf = ByteBuffer.allocate(varintBytes.size + payload.size)
        buf.put(varintBytes)
        buf.put(payload)
        return buf.array()
    }

    fun decodeDatagram(data: ByteArray): Pair<Long, ByteArray>? {
        if (data.isEmpty()) return null
        val (ctxId, varintLen) = decodeVarint(data) ?: return null
        val payload = data.copyOfRange(varintLen, data.size)
        return Pair(ctxId, payload)
    }

    private fun encodeVarint(value: Long): ByteArray {
        return if (value < 64) {
            byteArrayOf(value.toByte())
        } else if (value < 16384) {
            val v = (0x4000 or (value.toInt() and 0x3FFF))
            byteArrayOf((v shr 8).toByte(), (v and 0xFF).toByte())
        } else {
            val v = (0x80000000L or (value and 0x3FFFFFFFL))
            ByteBuffer.allocate(4).putInt(v.toInt()).array()
        }
    }

    private fun decodeVarint(data: ByteArray): Pair<Long, Int>? {
        if (data.isEmpty()) return null
        val first = data[0].toInt() and 0xFF
        val prefix = first shr 6
        return when (prefix) {
            0 -> Pair((first and 0x3F).toLong(), 1)
            1 -> {
                if (data.size < 2) null
                else {
                    val v = ((first and 0x3F) shl 8) or (data[1].toInt() and 0xFF)
                    Pair(v.toLong(), 2)
                }
            }
            2 -> {
                if (data.size < 4) null
                else {
                    val buf = ByteBuffer.wrap(data)
                    val v = buf.getInt().toLong() and 0x3FFFFFFFL
                    Pair(v, 4)
                }
            }
            else -> null
        }
    }
}
